### DISCUSSION TITLE
init.lua not being found by vscode-neovim?

### DISCUSSION THREAD

**Discussion by 1POP1k33kxx0x0xz** [vscode-neovim/vscode-neovim#2280] [Q&A] [answered]
My init.lua isn't being found by the extension on Windows 11 with neovim v0.10.2. My settings.json has the following lines:
```
    "vscode-neovim.neovimInitVimPaths.win32": "C:\\Users\\USER\\AppData\\Local\\nvim\\init.lua",
    "vscode-neovim.neovimExecutablePaths.win32": "C:\\Program Files\\Neovim\\bin\\nvim.exe"
```

Here are the logs at startup:
```
[info] MainController: Starting nvim: C:\Program Files\Neovim\bin\nvim.exe -N --embed --cmd source c:\\Users\\USER\\.vscode\\extensions\\asvetliakov.vscode-neovim-1.18.12/runtime/vscode-neovim.vim -u C:\Users\USER\AppData\Local\nvim\init.lua
2024-10-11 00:19:23.474 [info] MainController: Nvim info: {
  nvimVersion: {
    api_prerelease: false,
    api_compatible: 0,
    prerelease: false,
    minor: 10,
    major: 0,
    build: 'g8b9864200',
    patch: 2,
    api_level: 12
  },
  logFile: 'C:\\Users\\USER\\AppData\\Local\\nvim-data\\log',
  configDir: 'C:\\Users\\USER\\AppData\\Local\\nvim'
}
2024-10-11 00:19:23.504 [info] MessagesManager: ''
```

Both paths configured in settings.json are valid. The extension just doesn't seem to want to register my init.lua.

**Comment by justinmk** [comment]
Say specifically what you mean by "not found" or "not registered". The logs that you provided (thanks, that is helpful) clearly show `-u C:\Users\USER\AppData\Local\nvim\init.lua`, which means the configured init.lua is sent to Nvim.

From vscode, try `:echo $MYVIMRC` or `:edit $MYVIMRC` (to open the file) to confirm that Nvim found the file.

Most likely the file has an error, you can trying checking `:messages`. Or try `:source $MYVIMRC` and see if it outputs an error.

**Comment by 1POP1k33kxx0x0xz** [comment]
:edit $MYVIMRC seems to open an empty file. Surprisingly, in :messages the config outputs the "config loaded" message I put in my init file. While :source $MYVIMRC doesn't work, I can :source the file path of my init.lua file, and that seems to work. Not sure what is going on here.

**Reply by justinmk** [reply]
:messages shows no errors?

what does `:scriptnames` show?

**Reply by 1POP1k33kxx0x0xz** [reply]
Yes. Messages shows no errors. The only message shown is the config loaded message. Here is my scriptnames (look at line 23):

```
  1: c:\\Users\\USER\\.vscode\\extensions\\asvetliakov.vscode-neovim-1.18.12\runtime\vscode-neovim.vim
  2: C:\Program Files\Neovim\share/nvim/runtime/syntax/syntax.vim
  3: C:\Program Files\Neovim\share\nvim\runtime\syntax\synload.vim
  4: C:\Program Files\Neovim\share\nvim\runtime\filetype.lua
  5: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-code-actions.vim
  6: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-file-commands.vim
  7: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-jumplist.vim
  8: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-motion.vim
  9: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-scrolling.vim
 10: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-tab-commands.vim
 11: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\overrides\vscode-window-commands.vim
 12: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\clipboard.lua
 13: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\code_actions.lua
 14: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\filetype.lua
 15: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\inccommand.lua
 16: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\integrations.lua
 17: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\lsp\buf.lua
 18: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\statusline.lua
 19: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\treesitter.lua
 20: c:\\Users\USER\.vscode\extensions\asvetliakov.vscode-neovim-1.18.12\runtime\vscode\ui.lua
 21: C:\Program Files\Neovim\share\nvim\runtime\ftplugin.vim
 22: C:\Program Files\Neovim\share\nvim\runtime\indent.vim
 23: ~\AppData\Local\nvim\init.lua <-- The INIT file is apparently seen by the extension? (not sure if the ~ on windows is registered properly... maybe this is the issue?)
 24: C:\Program Files\Neovim\share\nvim\runtime\plugin\gzip.vim
 25: C:\Program Files\Neovim\share\nvim\runtime\plugin\matchit.vim
 26: C:\Program Files\Neovim\share\nvim\runtime\pack\dist\opt\matchit\plugin\matchit.vim
 27: C:\Program Files\Neovim\share\nvim\runtime\plugin\matchparen.vim
 28: C:\Program Files\Neovim\share\nvim\runtime\plugin\netrwPlugin.vim
 29: C:\Program Files\Neovim\share\nvim\runtime\plugin\rplugin.vim
 30: C:\Program Files\Neovim\share\nvim\runtime\plugin\shada.vim
 31: C:\Program Files\Neovim\share\nvim\runtime\plugin\spellfile.vim
 32: C:\Program Files\Neovim\share\nvim\runtime\plugin\tarPlugin.vim
 33: C:\Program Files\Neovim\share\nvim\runtime\plugin\tutor.vim
 34: C:\Program Files\Neovim\share\nvim\runtime\plugin\zipPlugin.vim
 35: C:\Program Files\Neovim\share\nvim\runtime\plugin\editorconfig.lua
 36: C:\Program Files\Neovim\share\nvim\runtime\plugin\man.lua
 37: C:\Program Files\Neovim\share\nvim\runtime\plugin\osc52.lua
 38: C:\Program Files\Neovim\share\nvim\runtime\plugin\tohtml.lua
```

So according to scriptnames and messages, the extension clearly sees the init.lua file, but I don't think it loads it properly. When I try doing :source <init path>, the configs load correctly (things like relative line numbers). Otherwise, no configs are applied on startup. FYI I did not put any ~ in any of the file paths in settings.json.

**Reply by justinmk** [reply]
> ~\AppData\Local\nvim\init.lua <-- The INIT file is apparently seen by the extension? (not sure if the ~ on windows is registered properly... maybe this is the issue?)

That is just a cosmetic/presentation difference. Your logs clearly showed that `-u C:\Users\USER\AppData\Local\nvim\init.lua` was passed, and Nvim was able to recognize that as your home dir.

**Reply by 1POP1k33kxx0x0xz** [reply]
So now what? My config has no issues, the extension just seems to not load it properly. Perhaps this is an actual bug I should post in issues?

**Reply by justinmk** [reply]
No one can help you until you provide more concrete details about what you mean by "init.lua was not registered". So far all evidence suggests that vscode-neovim passed it, and Nvim recognized it.

What specific things did you try? If you set an option at the top of your init.lua, does it get set? If you define a mapping in your init.lua, does `:map` show it? Things like that. 

If you call [vscode.notify()](https://github.com/vscode-neovim/vscode-neovim?tab=readme-ov-file#%EF%B8%8F-api) from your init.lua what happens?

Please debug yourself and provide specific, concise info.

**Reply by 1POP1k33kxx0x0xz** [reply]
I'm starting to think that the way init.lua works for vscode-neovim is different from neovim. I am using very basic configs that work perfectly in neovim with no issues. All of my keybinds are appear in :map, but while all work in regular neovim, only my <Leader> keybind works in vscode-neovim.

Keep in mind that running :source ~/AppData/Local/nvim/init.lua AFTER the config loads does resolve all these issues. I don't want to have to do that every time I open a new folder.

init.lua:
``` init.lua

if vim.g.vscode then
        local vscode = require("vscode")
        print = vscode.notify
end

local function mapPairs(maps, mapOpts) -- this works in neovim with no problems
        for _, mapArgs in ipairs(maps) do
                local charArray = {}

                for char in mapArgs[1]:gmatch("%a") do
                        table.insert(charArray, char)
                end

                vim.keymap.set(charArray, mapArgs[2], mapArgs[3], mapOpts)
        end
end

local noremaps = {
        {"n", "<C-u>", "<C-u>zz"}, -- normal mode, set ctrl-u to ctrl-u + zz
        {"n", "<C-d>", "<C-d>zz"},
        {"vinc", "<C-c>", "<Esc>"}, -- argument 1 (vinc) is for multiple modes: visual, insert, normal, command (supported by the function)
        {"n", "<Leader>r", ":!sync.bat<Cr>"} -- this is the only keybinding recognized by vscode-nvim. the rest are only recognized by nvim.
}

vim.g.mapleader = " "
vim.opt.number = true -- < both options do not work
vim.opt.relativenumber = true -- <
mapPairs(noremaps, {silent = true, noremap = true})

print("Config loaded") -- print (vscode.notify) works as expected
```

My config is very minimal atm, but I plan on adding more keybindings, hence the mapPairs function.

**Reply by justinmk** [reply]
> I'm starting to think that the way init.lua works for vscode-neovim is different from neovim

The only difference is that vscode-neovim includes its own configuration/plugins. It invokes `nvim` exactly as shown in your logs:

    C:\Program Files\Neovim\bin\nvim.exe -N --embed --cmd source c:\\Users\\USER\\.vscode\\extensions\\asvetliakov.vscode-neovim-1.18.12/runtime/vscode-neovim.vim -u C:\Users\USER\AppData\Local\nvim\init.lua

Did you try removing the `vscode-neovim.neovimInitVimPaths.win32` vscode setting? There is no reason to set that since you are using the default path expected by Nvim.

Oh, I just noticed something weird: in your logs, `source c:\\Users\\USER\\...` has escaped slashes but `-u C:\Users\USER\AppData\` does not. This looks like a bug, so I created an issue: https://github.com/vscode-neovim/vscode-neovim/issues/2285

As a workaround, you could use `/` slashes in your setting (or just drop the setting as I mentioned). There is never a reason to use `\` slashes on Windows.

**Reply by 1POP1k33kxx0x0xz** [reply]
Removing the vscode-neovim.neovimInitVimPaths.win32 does not fix my issue. I'm glad that you found a bug, but I still have the same exact issue with my init.lua, and it can't be vscode "not finding" init.lua like I originally stated. All of my configs work with no issue on the plain neovim.exe, but do not work in vscode. After removing the win32 setting, my "Config loaded" message still appears, which tells me that the file path escaping was not the issue, and instead, is an issue with the extension supporting the configs in init.lua.

**Reply by justinmk** [reply]
> After removing the win32 setting, my "Config loaded" message still appears, which tells me that the file path escaping was not the issue, and instead, is an issue with the extension supporting the configs in init.lua.

What in the world does this mean? The init was sourced and runs your code, yet you say "the extension does not support the configs".

**Reply by 1POP1k33kxx0x0xz** [reply]
I mean what I've been trying to say this entire time. The certain vim.opt and some keybindings aren't working with vscode-nvim. I've provided as much details as possible. I don't want to have to repeat things that were already stated.

**Comment by 1POP1k33kxx0x0xz** [comment]
Unsolved. My best guess is that this is a bug with vscode-neovim that treats init.lua files differently from regular nvim. Certain page up/page down shortcuts don't work correctly even when running the source <init file path>. Relative line numbers do not appear unless you run the source command, and certain configs in init.lua work without needing to manually be sourced with the source command.

My ad hoc solution is running the source command (doesn't completely fix) or just not using the extension altogether. I find it to be noticeably slower than regular nvim with a lot of things.

**Comment by justinmk** [comment, marked as answer]
> `:edit $MYVIMRC` seems to open an empty file

That's just normal Vim/Nvim behavior. `$MYVIMRC` is not set when you specify the config via `-u`.

> The certain vim.opt and some keybindings aren't working with vscode-nvim.

That is because vscode-neovim sets those options per-buffer: https://github.com/vscode-neovim/vscode-neovim/blob/3e343d683920de1c2e0c264207fb845b4783626d/runtime/lua/vscode/sync-options.lua#L73-L78

As a workaround, you could set `"editor.lineNumbers": "relative"` in vscode settings (not Nvim).
